test/gtest/common/test.h

```diff
@@ -0,0 +1,250 @@
+
+    static void *thread_func(void *arg);
+
+    pthread_barrier_t    m_barrier;
+};
+
+#define UCC_TEST_BASE_IMPL \
+    virtual void SetUp() { \
+        test_base::SetUpProxy(); \
+    } \
+    \
+    virtual void TearDown() { \
+        test_base::TearDownProxy(); \
+    } \
+    virtual void TestBody() { \
+        test_base::TestBodyProxy(); \
+    }
+
+/*
+ * Base class from generic tests
+ */
+class test : public testing::Test, public test_base {
+public:
+    UCC_TEST_BASE_IMPL;
+};
+
+/*
+ * Base class from generic tests with user-defined parameter
+ */
+template <typename T>
+class test_with_param : public testing::TestWithParam<T>, public test_base {
+public:
+    UCC_TEST_BASE_IMPL;
+};
+
+}
+
+/*
+ * Helper macro
+ */
+#define UCC_TEST_(test_case_name, test_name, parent_id, \
+                  num_threads, skip_cond, skip_reason, ...) \
+class GTEST_TEST_CLASS_NAME_(test_case_name, test_name) : public test_case_name { \
+ public: \
+  GTEST_TEST_CLASS_NAME_(test_case_name, test_name)() { \
+  } \
+ protected: \
+  virtual void init() { \
+      test_case_name::init(); \
+  } \
+ private: \
+  virtual void check_skip_test() { \
+      if (skip_cond) { \
+          UCC_TEST_SKIP_R(skip_reason); \
+      } \
+  } \
+  virtual void test_body(); \
+  static ::testing::TestInfo* const test_info_;\
+  GTEST_DISALLOW_COPY_AND_ASSIGN_(\
+      GTEST_TEST_CLASS_NAME_(test_case_name, test_name));\
+}; \
+\
+::testing::TestInfo* const GTEST_TEST_CLASS_NAME_(test_case_name, test_name)\
+  ::test_info_ = \
+    ::testing::internal::MakeAndRegisterTestInfo( \
+        #test_case_name, \
+        (num_threads == 1) ? #test_name : #test_name "/mt_" #num_threads, \
+        "", "", \
+        (parent_id), \
+        test_case_name::SetUpTestCase, \
+        test_case_name::TearDownTestCase, \
+        new ::testing::internal::TestFactoryImpl< \
+            GTEST_TEST_CLASS_NAME_(test_case_name, test_name)>); \
+void GTEST_TEST_CLASS_NAME_(test_case_name, test_name)::test_body()
+
+
+/*
+ * Define test fixture with modified configuration
+ */
+#define UCC_TEST_F(test_fixture, test_name, ...) \
+  UCC_TEST_(test_fixture, test_name, \
+            ::testing::internal::GetTypeId<test_fixture>(), \
+            1, 0, "", __VA_ARGS__)
+
+
+/*
+ * Define test fixture with modified configuration and check skip condition
+ */
+#define UCC_TEST_SKIP_COND_F(test_fixture, test_name, skip_cond, ...) \
+  UCC_TEST_(test_fixture, test_name, \
+            ::testing::internal::GetTypeId<test_fixture>(), \
+            1, skip_cond, #skip_cond, __VA_ARGS__)
+
+
+/*
+ * Define test fixture with multiple threads
+ */
+#define UCC_MT_TEST_F(test_fixture, test_name, num_threads, ...) \
+  UCC_TEST_(test_fixture, test_name, \
+            ::testing::internal::GetTypeId<test_fixture>(), \
+            num_threads, 0, "", __VA_ARGS__)
+
+
+/*
+ * Helper macro
+ */
+#define UCC_TEST_P_(test_case_name, test_name, num_threads, \
```

Comment:
What is the difference between UCC_TEST_F and UCC_TEST_P? When should i use either of those?

Reply:
If you're still wondering about this, and if this is similar to UCX, then UCC_TEST_P is the parameterized version of UCC_TEST_F. You can use UCC_TEST_P when you want to run the exact same test with different values of an input parameter.
The value of the parameter can be accessed by `GetParam()` macro in the body of UCC_TEST_P. At the end of your test file, you should use one of `INSTANTIATE_TEST_CASE_P()` macros which instantiate all parameterized tests with a given list of parameter values that are passed to `INSTANTIATE_TEST_CASE_P()`.
Hope this helps.

Reply:
thanks!